From 0aa46f9e6689af070161f3a5855150e7d9be63fd Mon Sep 17 00:00:00 2001
From: ziqic <ziqic@qti.qualcomm.com>
Date: Mon, 24 Mar 2025 17:38:46 +0800
Subject: [PATCH] PENDING: defer no-map memory init process

Defer no-map memory init process and run it parallelly
later to optimize bootkpi.

Change-Id: I25d0f50c83abb9cf3ccc92fc2e45d2dff15c1d07
Signed-off-by: ziqic <ziqic@qti.qualcomm.com>
---
 arch/arm64/Kconfig |  1 +
 mm/Kconfig         |  3 +++
 mm/memblock.c      |  5 +++++
 mm/mm_init.c       | 49 ++++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 58 insertions(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 658c6a61ab6f..7e9a816f0284 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -256,6 +256,7 @@ config ARM64
 	select TRACE_IRQFLAGS_SUPPORT
 	select TRACE_IRQFLAGS_NMI_SUPPORT
 	select HAVE_SOFTIRQ_ON_OWN_STACK
+	select DEFER_NO_MAP_MEM_PAGE_INIT
 	help
 	  ARM 64-bit (AArch64) Linux support.
 
diff --git a/mm/Kconfig b/mm/Kconfig
index c11cd01169e8..774d499e5b9d 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -1007,6 +1007,9 @@ config DEFERRED_STRUCT_PAGE_INIT
 	  lifetime of the system until these kthreads finish the
 	  initialisation.
 
+config DEFER_NO_MAP_MEM_PAGE_INIT
+	bool
+
 config PAGE_IDLE_FLAG
 	bool
 	select PAGE_EXTENSION if !64BIT
diff --git a/mm/memblock.c b/mm/memblock.c
index 047dce35cf6e..9358a8a06408 100644
--- a/mm/memblock.c
+++ b/mm/memblock.c
@@ -2122,14 +2122,18 @@ static void __init memmap_init_reserved_pages(void)
 	struct memblock_region *region;
 	phys_addr_t start, end;
 	int nid;
+
+#ifndef CONFIG_DEFER_NO_MAP_MEM_PAGE_INIT
 	unsigned long max_reserved;
 
 	/*
 	 * set nid on all reserved pages and also treat struct
 	 * pages for the NOMAP regions as PageReserved
 	 */
+
 repeat:
 	max_reserved = memblock.reserved.max;
+
 	for_each_mem_region(region) {
 		nid = memblock_get_region_node(region);
 		start = region->base;
@@ -2147,6 +2151,7 @@ static void __init memmap_init_reserved_pages(void)
 	 */
 	if (max_reserved != memblock.reserved.max)
 		goto repeat;
+#endif
 
 	/* initialize struct pages for the reserved regions */
 	for_each_reserved_mem_region(region) {
diff --git a/mm/mm_init.c b/mm/mm_init.c
index 77fd04c83d04..83c458d86462 100644
--- a/mm/mm_init.c
+++ b/mm/mm_init.c
@@ -2378,6 +2378,46 @@ void set_zone_contiguous(struct zone *zone)
 	zone->contiguous = true;
 }
 
+#ifdef CONFIG_DEFER_NO_MAP_MEM_PAGE_INIT
+static int memmap_init_reserved_pages_thread(void *args)
+{
+	struct memblock_region *region;
+	phys_addr_t start, end;
+	int nid;
+	unsigned long max_reserved;
+
+	/*
+	 * set nid on all reserved pages and also treat struct
+	 * pages for the NOMAP regions as PageReserved
+	 */
+
+repeat:
+	max_reserved = memblock.reserved.max;
+
+	for_each_mem_region(region) {
+		nid = memblock_get_region_node(region);
+		start = region->base;
+		end = start + region->size;
+
+		if (memblock_is_nomap(region))
+			reserve_bootmem_region(start, end, nid);
+
+		memblock_set_node(start, region->size, &memblock.reserved, nid);
+	}
+	/*
+	 * 'max' is changed means memblock.reserved has been doubled its
+	 * array, which may result a new reserved region before current
+	 * 'start'. Now we should repeat the procedure to set its node id.
+	 */
+	if (max_reserved != memblock.reserved.max)
+		goto repeat;
+
+	pr_err("memmap_init_reserved_pages kthread finish.\n");
+
+	return 0;
+}
+#endif
+
 void __init page_alloc_init_late(void)
 {
 	struct zone *zone;
@@ -2404,6 +2444,15 @@ void __init page_alloc_init_late(void)
 	files_maxfiles_init();
 #endif
 
+#ifdef CONFIG_DEFER_NO_MAP_MEM_PAGE_INIT
+	struct task_struct *t;
+	t = kthread_run(memmap_init_reserved_pages_thread, NULL, "reserved-memory");
+	if (IS_ERR(t)) {
+		pr_err("Failed to create memmap_init_reserved_pages thread.\n");
+	}
+
+#endif
+
 	buffer_init();
 
 	/* Discard memblock private memory */
-- 
2.34.1

